pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -831,3 +836,59 @@ func (h *upgradeHandler) startVM(ctx context.Context, vm *kubevirtv1.VirtualMach
 		Do(ctx)
 	return res.Error()
 }
+
+// checkLogReadyCondition times out LogReady condition, thus moving on to the next step if upgradeLogCondLogReadyTimeout
+// time has passed since UpgradeLog's logging infrastructure setup started
+func (h *upgradeHandler) checkLogReadyCondition(upgrade *harvesterv1.Upgrade) (*harvesterv1.Upgrade, error) {
+	upgradeConfig, err := settings.DecodeConfig[settings.UpgradeConfig](settings.UpgradeConfigSet.Get())
+	if err != nil {
+		logrus.Errorf("Failed to get UpgradeConfig")
+		return upgrade, err
+	}
+
+	timeoutStr := upgradeConfig.LogReadyTimeout
+	timeout, err := strconv.Atoi(timeoutStr)
+	if err != nil {
+		return upgrade, fmt.Errorf("invalid value for image preload timeout: %s", timeoutStr)
+	}
+	timeoutDuration := time.Duration(timeout) * time.Minute
+
+	if timeoutStr == "" || timeoutDuration < 1*time.Minute || timeoutDuration > 20*time.Minute {
+		logrus.Warnf("invalid logReadyTimeout must be between 1 to 20 minutes, given: %s", timeoutStr)
+		logrus.Infof("Using default timeoutStr: %v", settings.UpgradeLogReadyTimeoutDefault)
+		timeoutDuration = settings.UpgradeLogReadyTimeoutDefault
+	}
+
+	ts := harvesterv1.LogReady.GetLastUpdated(upgrade)
+	t, err := time.Parse(time.RFC3339, ts)
+	if err != nil {
+		logrus.Errorf("Failed to parse last updated time: %v", err)
+		return upgrade, nil
+	}
+	if time.Since(t) >= timeoutDuration {
+		toUpdate := upgrade.DeepCopy()
+		upgradeLog, err := h.upgradeLogCache.Get(util.HarvesterSystemNamespaceName, upgrade.Status.UpgradeLog)
+		if err != nil {
+			logrus.Errorf("Failed to get upgrade log: %v", err)
+			setLogReadyCondition(toUpdate, corev1.ConditionFalse, "UpgradeLogGetFailed", "Failed to get upgrade log")
+			return h.upgradeClient.Update(toUpdate)
+		}
```

Comment:
This doesn't feel like the right thing to do for the situation. Can you PTAL and suggest any changes/improvement? Should we also do `setUpgradeCompletedCondition`? I could be totally wrong, but wonder if this will go into a loop of trying to get upgradeLog.

Reply:
As you have already waited until the timeout, even if unluckily the upgradeLog was deleted by user or due to other reason, you can't get it, it is ok to fail the upgrade.